adcc/kernel/modified_transition_moments.py

```diff
@@ -0,0 +1,130 @@
+## it under the terms of the GNU General Public License as published
+## by the Free Software Foundation, either version 3 of the License, or
+## (at your option) any later version.
+##
+## adcc is distributed in the hope that it will be useful,
+## but WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with adcc. If not, see <http://www.gnu.org/licenses/>.
+##
+## ---------------------------------------------------------------------
+from math import sqrt
+
+from adcc import block as b
+from adcc.AdcMethod import AdcMethod
+from adcc.functions import einsum, evaluate
+from adcc.AmplitudeVector import AmplitudeVector
+
+import libadcc
+
+
+def modified_transition_moments_adc0(mp, dipop, intermediates):
+    return AmplitudeVector(dipop[b.ov])
+
+
+def modified_transition_moments_adc1(mp, dipop, intermediates):
+    f1 = dipop[b.ov] - einsum("ijab,jb->ia", mp.t2(b.oovv), dipop[b.ov])
+    return AmplitudeVector(f1)
+
+
+def modified_transition_moments_adc2(mp, dipop, intermediates):
+    t2 = mp.t2(b.oovv)
+    p0 = mp.mp2_diffdm
+
+    f1 = (
+        + dipop[b.ov]
+        - einsum("ijab,jb->ia", t2,
+                 + dipop[b.ov] - 0.5 * einsum("jkbc,kc->jb", t2, dipop[b.ov]))
+        + 0.5 * einsum("ij,ja->ia", p0[b.oo], dipop[b.ov])
+        - 0.5 * einsum("ib,ab->ia", dipop[b.ov], p0[b.vv])
+        + einsum("ib,ab->ia", p0[b.ov], dipop[b.vv])
+        - einsum("ij,ja->ia", dipop[b.oo], p0[b.ov])
+        - einsum("ijab,jb->ia", mp.td2(b.oovv), dipop[b.ov])
+    )
+    f2 = (
+        + einsum("ijac,bc->ijab", t2, dipop[b.vv]).antisymmetrise(2, 3)
+        - einsum("ik,kjab->ijab", dipop[b.oo], t2).antisymmetrise(0, 1)
+    )
+    return AmplitudeVector(f1, f2)
+
+
+def modified_transition_moments_cvs_adc0(mp, dipop, intermediates):
+    return AmplitudeVector(dipop[b.cv])
+
+
+def modified_transition_moments_cvs_adc2(mp, dipop, intermediates):
+    f1 = (
+        + dipop[b.cv]
+        - einsum("Ib,ba->Ia", dipop[b.cv], intermediates.cv_p_vv)
+        - einsum("Ij,ja->Ia", dipop[b.co], intermediates.cv_p_ov)
+    )
+    f2 = (1 / sqrt(2)) * einsum("Ik,kjab->jIab", dipop[b.co], mp.t2(b.oovv))
+    return AmplitudeVector(f1, f2)
+
+
+DISPATCH = {
+    "adc0": modified_transition_moments_adc0,
+    "adc1": modified_transition_moments_adc1,
+    "adc2": modified_transition_moments_adc2,
+    "cvs-adc0": modified_transition_moments_cvs_adc0,
+    "cvs-adc1": modified_transition_moments_cvs_adc0,  # Identical to CVS-ADC(0)
+    "cvs-adc2": modified_transition_moments_cvs_adc2,
+}
+
+
+def modified_transition_moments(method, ground_state, dipole_operator=None,
+                                intermediates=None):
+    """Compute the modified transition moments (MTM) for the provided
+    ADC method with reference to the passed ground state.
```

Comment:
Just out of interest... is there a possibility to easily add LaTex equations to the documentation?
Then we could have $F_J(\op{d}) = \ISB{J}\op{d}\ket{\Psi_0}$ or something of the sort 😄

Reply:
Sure! It's processed by sphinx so you can just use RST like in https://raw.githubusercontent.com/adc-connect/adcc/master/docs/theory.rst. We have to be careful with this to not blow up the docstrings too much. I think a general equation in the dispatch function is fine, but I would be reluctant to document the method-specific implementations to save space.

Reply:
Yeah, I was not talking about method-specific stuff. This would be insane. Just general equations, but I would not do it in this PR.

Reply:
Agree.